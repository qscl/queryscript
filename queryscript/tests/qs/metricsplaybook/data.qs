let src_contract_stream = load('raw_csvs/src_contract_stream.csv');
let src_dim_customer = load('raw_csvs/src_dim_customer.csv');

mat contract_stream =
	select
		id,
		customer_id,
		activity,
		plan_type,
		timestamp,
		row_number() over(partition by customer_id, activity order by timestamp asc) as activity_occurrence,
		-- XXX: Need to support lead()
		-- lead(timestamp, 1) over(partition by customer_id, activity order by timestamp asc) as activity_repeated_at
		timestamp AS activity_repeated_at
	from
		src_contract_stream;


mat dim_customer =
	select *,
	-- XXX need to support substring, date_trunc, etc.
	-- substring(date_trunc('month', first_contract_signed_date)::text, 1, 7) as cohort
	'todo' as cohort
	from src_dim_customer;


SELECT * FROM contract_stream ORDER BY id LIMIT 5;
SELECT * FROM dim_customer ORDER BY id LIMIT 5;
