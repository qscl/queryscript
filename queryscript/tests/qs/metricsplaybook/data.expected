{
    "compile_errors": [],
    "decls": {
        "let contract_stream": [{
        	id Int64,
        	customer_id Int64,
        	activity Utf8,
        	plan_type Utf8,
        	timestamp Utf8,
        	activity_occurrence Int64,
        	activity_repeated_at Utf8,
        }],
        "let dim_customer": [{
        	id Int64,
        	segment Utf8,
        	channel Utf8,
        	first_contract_signed_date Utf8,
        	cohort Utf8,
        }],
        "let src_contract_stream": External<[{
        	id Int64,
        	customer_id Int64,
        	timestamp Utf8,
        	activity Utf8,
        	revenue_impact Float64,
        	plan_type Utf8,
        	activity_occurrence Int64,
        	activity_repeated_at Utf8,
        }]>,
        "let src_dim_customer": External<[{
        	id Int64,
        	segment Utf8,
        	channel Utf8,
        	first_contract_signed_date Utf8,
        }]>,
    },
    "queries": [
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "id",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "customer_id",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "activity",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "plan_type",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "timestamp",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "activity_occurrence",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "activity_repeated_at",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| id | customer_id | activity                     | plan_type | timestamp            | activity_occurrence | activity_repeated_at |\n|----|-------------|------------------------------|-----------|----------------------|---------------------|----------------------|\n| 1  | 156         | customer_churn_committed     | plan4     | 2022-05-23T08:28:22Z | 1                   | 2022-05-23T08:28:22Z |\n| 2  | 258         | customer_churn_committed     | plan3     | 2022-05-10T13:15:51Z | 1                   | 2022-05-10T13:15:51Z |\n| 3  | 296         | contraction_contract_started | plan4     | 2022-02-15T14:03:09Z | 1                   | 2022-02-15T14:03:09Z |\n| 4  | 215         | expansion_contract_started   | plan4     | 2022-06-05T22:31:43Z | 1                   | 2022-06-05T22:31:43Z |\n| 5  | 284         | new_contract_started         | plan4     | 2022-09-03T20:53:59Z | 2                   | 2022-09-03T20:53:59Z |",
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "id",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "segment",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "channel",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "first_contract_signed_date",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "cohort",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| id | segment  | channel  | first_contract_signed_date | cohort |\n|----|----------|----------|----------------------------|--------|\n| 1  | segment2 | channel4 | 2022-05-20T11:24:40Z       | todo   |\n| 2  | segment7 | channel1 | 2022-03-13T09:36:31Z       | todo   |\n| 3  | segment3 | channel3 | 2021-12-24T10:37:19Z       | todo   |\n| 4  | segment3 | channel4 | 2022-09-26T22:04:35Z       | todo   |\n| 5  | segment7 | channel3 | 2022-01-21T20:27:11Z       | todo   |",
            },
        ),
    ],
}